x/btcstaking/keeper/power_dist_change.go

```diff
@@ -91,11 +91,17 @@ func (k Keeper) recordVotingPowerAndCache(ctx context.Context, prevDc, newDc *ty
 
 	// label fps with whether it has timestamped pub rand so that these fps
 	// will not be assigned voting power
-	for _, fp := range newDc.FinalityProviders {
+	for _, fpDistInfo := range newDc.FinalityProviders {
 		// TODO calling HasTimestampedPubRand potentially iterates
-		// all the pub rand committed by the fp, which might slow down
+		// all the pub rand committed by the fpDistInfo, which might slow down
 		// the process, need optimization
-		fp.IsTimestamped = k.FinalityKeeper.HasTimestampedPubRand(ctx, fp.BtcPk, babylonTipHeight)
+		fpDistInfo.IsTimestamped = k.FinalityKeeper.HasTimestampedPubRand(ctx, fpDistInfo.BtcPk, babylonTipHeight)
+
+		fp, err := k.GetFinalityProvider(ctx, fpDistInfo.BtcPk.MustMarshal())
+		if err != nil {
+			panic(fmt.Errorf("the finality provider %s does not exist: %w", fp.BtcPk.MarshalHex(), err))
+		}
+		fpDistInfo.IsJailed = fp.IsJailed()
```

Comment:
This will cause some overhead imo -- upon every new block computing the new voting power table will need to retrieve all FPs in the DB, leading to `O(n)` overhead in DB read where `n` is the number of FPs. This was not the case where the DB overhead is `O(# of BTC delegations with state updates)`.

Since we will have `MsgUnjail` anyway, how about we do the state update using the event-based approach similar to the existing events? That is,

- Add a field `IsJailed` to `FinalityProviderDistInfo`
- Upon jailing a FP, push a jailed event to the event queue
- Upon `MsgUnjail` is approved, push an unjailed event to the event queue
- `ProcessAllPowerDistUpdateEvents` processes all the corresponding jailed/unjailed events and sets `IsJailed` for corresponding `FinalityProviderDistInfo` objects

Reply:
Nice point! Chaned to setting `IsJailed` to `FinalityProviderDistInfo` upon event processing and will have `UnjailEvent` to revert the flag in the subsequent PR